modules/etcd/etcd.go

```diff
@@ -0,0 +1,241 @@
+package etcd
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/testcontainers/testcontainers-go"
+	tcnetwork "github.com/testcontainers/testcontainers-go/network"
+)
+
+const (
+	ClientPort          = "2379"
+	PeerPort            = "2380"
+	DataDir             = "/data.etcd"
+	DefaultClusterToken = "mys3cr3ttok3n"
+	scheme              = "http"
+)
+
+// EtcdContainer represents the etcd container type used in the module
+type EtcdContainer struct {
+	testcontainers.Container
+	ClusterToken string
+	Nodes        []*EtcdContainer
+}
+
+// Run creates an instance of the etcd container type
+func Run(ctx context.Context, img string, opts ...testcontainers.ContainerCustomizer) (*EtcdContainer, error) {
+	req := testcontainers.ContainerRequest{
+		Image: img,
+		Cmd:   []string{},
+	}
+
+	genericContainerReq := testcontainers.GenericContainerRequest{
+		ContainerRequest: req,
+		Started:          true,
+	}
+
+	settings := defaultOptions(&req)
+	for _, opt := range opts {
+		if apply, ok := opt.(Option); ok {
+			apply(&settings)
+		}
+		if err := opt.Customize(&genericContainerReq); err != nil {
+			return nil, err
+		}
+	}
+
+	clusterOpts, err := configureCluster(ctx, &settings, opts)
+	if err != nil {
+		return nil, fmt.Errorf("configure cluster: %w", err)
+	}
+
+	// configure CMD with the nodes
+	genericContainerReq.Cmd = configureCMD(settings)
+
+	if settings.clusterNetwork != nil {
+		// apply the network to the current node
+		err := tcnetwork.WithNetwork([]string{settings.Nodes[settings.currentNode]}, settings.clusterNetwork)(&genericContainerReq)
+		if err != nil {
+			return nil, fmt.Errorf("with network: %w", err)
+		}
+	}
+
+	container, err := testcontainers.GenericContainer(ctx, genericContainerReq)
+	var c *EtcdContainer
+	if container != nil {
+		c = &EtcdContainer{Container: container}
+	}
+
+	if err != nil {
+		return c, fmt.Errorf("generic container: %w", err)
+	}
+
+	// only the first node creates the cluster
+	if settings.currentNode == 0 {
+		for i := 1; i < len(settings.Nodes); i++ {
+			// move to the next node
+			clusterNode, err := Run(ctx, req.Image, append(clusterOpts, withCurrentNode(i))...)
+			if err != nil {
+				// return the parent cluster node and the error, so the caller can clean up.
+				return c, fmt.Errorf("run cluster node: %w", err)
+			}
+
+			c.Nodes = append(c.Nodes, clusterNode)
```

Comment:
suggestion: `clusterNode` could also be valid here, requiring clean up.
```suggestion
			clusterNode, err := Run(ctx, req.Image, append(clusterOpts, withCurrentNode(i))...)
			if clusterNode != nil {
				c.Nodes = append(c.Nodes, clusterNode)
			}
			if err != nil {
				// return the parent cluster node and the error, so the caller can clean up.
				return c, fmt.Errorf("run cluster node: %w", err)
			}
```

Reply:
Originally, the Nodes field represents all the child nodes, not all cluster nodes, therefore the initial node is not part of that list. I added tests for describing those use cases.

Maybe worth it to rename that field to childNodes (private)?